File: medium/41.cpp
#include <iostream>
#include <vector>
#include <functional>
#include <algorithm>    // std::min_element, std::max_element
#include <initializer_list>
#include <unordered_set>
#include <limits.h>

using namespace std;

class Solution {
public:
    bool isInSet(long e, std::unordered_set<int>& set) {
        return not(set.find(e) == set.end());
    }
    int firstMissingPositive(vector<int>& nums) {
        std::unordered_set<int> set (nums.begin(), nums.end());
        long ans = INT_MAX;
        while (!set.empty()){
            long elem = *set.begin();
            if (elem <= 0) {
                set.erase(elem);
                continue;
            }
            long e  = elem;
            while (isInSet(e - 1, set)){ set.erase(e); e--;
            }
            ans = min(e, ans);
            set.erase(e);
            e= elem + 1;
            while (isInSet(e, set)){ set.erase(e); e++;
            }
        }
        return ans;
    }
};

int main() {
    // vector<int> nums = {1, 2, 0};
    vector<int> nums = {3,4,-1,1};
    Solution sol = Solution();
    cout << sol.firstMissingPositive(nums) << endl;
    return 0;
}

File: HuaHua/41.py
from __future__ import annotations
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        i = 0
        while i < len(nums):
            n = nums[i]
            if n - 1 > len(nums) - 1 or n - 1 < 0 or n == i + 1 or nums[n - 1] == n:
                i += 1
            else:
                nums[i], nums[n - 1] = nums[n - 1], nums[i]
        for i, n in enumerate(nums):
            if n != i + 1:
                return i + 1
        return len(nums) + 1

if __name__ == '__main__':
    sol = Solution()
    nums = [1,2,0]
    print(sol.firstMissingPositive(nums))
    nums = [3, 4, -1, 1]
    print(sol.firstMissingPositive(nums))
    nums = [7,8,9,11,12]
    print(sol.firstMissingPositive(nums))
    nums = []
    print(sol.firstMissingPositive(nums))
    nums = [1]
    print(sol.firstMissingPositive(nums))
    nums = [1, 1]
    print(sol.firstMissingPositive(nums))
